File: frontend/src/types/order.ts
export type OrderStatus =
  | 'received'
  | 'confirmed'
  | 'preparing'
  | 'ready'
  | 'collected'
  | 'cancelled';

export type MenuCategory = string;

export interface MenuItem {
  readonly id: string;
  readonly venueId: string;
  readonly venue_id?: string;
  readonly name: string;
  readonly description: string | null;
  readonly price: number;
  readonly category: MenuCategory;
  readonly dietaryTags: string[];
  readonly dietary_tags?: string[];
  readonly imageUrl: string | null;
  readonly image_url?: string | null;
  readonly prepTimeMinutes: number;
  readonly prep_time_minutes?: number;
  readonly isAvailable: boolean;
  readonly is_available?: boolean;
  readonly isSoldOut: boolean;
  readonly is_sold_out?: boolean;
}

export interface CartItem {
  menuItem: MenuItem;
  menu_item?: MenuItem;
  quantity: number;
  specialInstructions: string;
  special_instructions?: string;
  price?: number;
  name?: string;
  menu_item_id?: string;
}

export interface OrderItemDetail {
  readonly menuItemId: string;
  readonly menu_item_id?: string;
  readonly name: string;
  readonly quantity: number;
  readonly unitPrice: number;
  readonly unit_price?: number;
  readonly subtotal: number;
  readonly specialInstructions: string | null;
  readonly special_instructions?: string | null;
}

export interface Order {
  readonly id: string;
  readonly orderCode: string;
  readonly order_code?: string;
  readonly userId: string;
  readonly user_id?: string;
  readonly venueId: string;
  readonly venue_id?: string;
  readonly items: OrderItemDetail[];
  readonly totalPrice: number;
  readonly total_amount?: number;
  readonly total_price?: number;
  readonly status: OrderStatus;
  readonly pickupZoneId: string | null;
  readonly pickup_zone_id?: string | null;
  readonly pickupSlot: string | null;
  readonly pickup_slot?: string | null;
  readonly specialInstructions: string | null;
  readonly special_instructions?: string | null;
  readonly isDemo: boolean;
  readonly is_demo?: boolean;
  readonly createdAt: string;
  readonly created_at?: string;
  readonly updatedAt: string | null;
  readonly updated_at?: string | null;
}

export const ORDER_STATUS_STEPS: OrderStatus[] = [
  'received',
  'confirmed',
  'preparing',
  'ready',
  'collected',
];

export interface OrderStatusState {
  order_code?: string;
  pickup_zone_name?: string;
  pickup_zone_id?: string;
  pickup_slot?: string | null;
  status: OrderStatus | Uppercase<OrderStatus>;
  updated_at?: string | null;
  items?: OrderItemDetail[];
  total_amount?: number;
}

File: backend/app/schemas/order.py
from pydantic import BaseModel, ConfigDict, Field, field_serializer, field_validator
from datetime import datetime
import uuid


class MenuItemResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    venue_id: uuid.UUID
    name: str
    description: str | None = None
    price: float
    category: str
    dietary_tags: list[str] = Field(default_factory=list)
    image_url: str | None = None
    prep_time_minutes: int
    is_available: bool
    is_sold_out: bool

    @field_serializer("category")
    def serialize_category(self, value: str) -> str:
        return str(value.value if hasattr(value, "value") else value).lower()


class OrderItemRequest(BaseModel):
    menu_item_id: uuid.UUID
    quantity: int = Field(ge=1, le=20)
    special_instructions: str | None = Field(default=None, max_length=200)


class CreateOrderRequest(BaseModel):
    items: list[OrderItemRequest] = Field(min_length=1, max_length=20)
    pickup_zone_id: uuid.UUID
    pickup_slot: datetime
    special_instructions: str | None = Field(default=None, max_length=500)
    is_demo: bool = Field(default=False)

    @field_validator("pickup_slot")
    @classmethod
    def validate_future_slot(cls, v: datetime) -> datetime:
        from datetime import timezone

        if v.astimezone(timezone.utc) < datetime.now(timezone.utc):
            raise ValueError("Pickup slot must be in the future")
        return v


class OrderResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    order_code: str
    user_id: uuid.UUID
    venue_id: uuid.UUID
    items: list[dict[str, object]]
    total_price: float
    status: str
    pickup_zone_id: uuid.UUID | None = None
    pickup_slot: datetime | None = None
    special_instructions: str | None = None
    is_demo: bool
    created_at: datetime
    updated_at: datetime | None = None

    @field_serializer("status")
    def serialize_status(self, value: str) -> str:
        return str(value.value if hasattr(value, "value") else value).lower()


class UpdateOrderStatusRequest(BaseModel):
    status: str

    @field_validator("status")
    @classmethod
    def validate_status(cls, v: str) -> str:
        normalized = v.strip().lower()
        valid = ["confirmed", "preparing", "ready", "collected", "cancelled"]
        if normalized not in valid:
            raise ValueError(f"Status must be one of {valid}")
        return normalized.upper()


class MenuAvailabilityRequest(BaseModel):
    is_sold_out: bool
    is_available: bool | None = None
